File: SafetyProgram.Core/ApplicationKernel.cs
using System;
using System.Collections.ObjectModel;
using SafetyProgram.Base;
using SafetyProgram.Core.IO;
using SafetyProgram.Core.Models;
using SafetyProgram.Core.Models.Factories;
using SafetyProgram.Core.Models.Serialization;

namespace SafetyProgram.Core
{
    /// <summary>
    /// Defines a standard implementation of IApplicationKernel
    /// </summary>
    public sealed class ApplicationKernel : IApplicationKernel
    {
        /// <summary>
        /// Construct an instance of an application kernel.
        /// </summary>
        public ApplicationKernel(IApplicationConfiguration configuration)
        {
            this.Configuration = configuration;

            document = new Document(
                new ObservableCollection<IDocumentObject>(),
                new A4Format());

            service = new InteractiveLocalFileService<IDocument>(
                new DocumentFactory(),
                new DocumentXml());
        }

        private IDocument document;

        /// <summary>
        /// Get or set the document open in the application, may be null.
        /// </summary>
        public IDocument Document
        {
            get
            {
                return document;
            }
            set
            {
                document = value;
                DocumentChanged.Raise(this, document);
            }
        }

        /// <summary>
        /// Occurs when the document in the appplication changes.
        /// </summary>
        public event EventHandler<
            GenericPropertyChangedEventArg<
                IDocument>> DocumentChanged;

        private IIOService<IDocument> service;

        /// <summary>
        /// Get or set the service used by the application.
        /// </summary>
        public IIOService<IDocument> Service
        {
            get
            {
                return service;
            }
            set
            {
                Helpers.NullCheck(service);
                service = value;
                ServiceChanged.Raise(this, service);
            }
        }

        /// <summary>
        /// Occurs when the IO service used by the application changes.
        /// </summary>
        public event EventHandler<
            GenericPropertyChangedEventArg<
                IIOService<IDocument>>> ServiceChanged;

        /// <summary>
        /// Get the configuration used by the application.
        /// </summary>
        public IApplicationConfiguration Configuration { get; private set; }
    }
}

File: SafetyProgram.Core/ApplicationKernel.fs
module SafetyProgram.Core

open System
open System.ComponentModel
open SafetyProgram.Core.Models
open SafetyProgram.Core.IO.Services
open SafetyProgram.Core.IO.LocalSvc
open SafetyProgram.Core.Models.Serialization.Core
open SafetyProgram.Core.Models.Serialization.DocumentXml
open System.Xml.Linq
open System.IO

// Data open in the application (could be local, could be databased).
type DataType = 
| LocalFile of FileStream
| BufferedFile

// Application state data.
type KernelData = {
    // The current document open in the application.
    Content : Option<Document * DataType>

    // The IO service used by the application.
    Service : DataService<Document>

    // The configuration used by the application.
    Configuration : ApplicationConfiguration
}

// Message passed back when attempting to modify the kernel.
type UpdateResult = Ok | Error of string

// Active service for monitoring and updating the application kernel.
type DataService<'a> = {
    Current : unit -> Async<'a>

    KernelDataChanged : IEvent<'a>

    Modify : ('a -> 'a) -> Async<UpdateResult>
}

/// Build a kernel service for the application kernel.
let buildKernelService<'a> (init : 'a) = 

    // Mutable: set the kernel data.
    let current = ref init

    // Track kernel data changes.
    let kernelDataChanged = new Event<'a>()

    // Serves incoming requests for the current kernel data.
    let currentData : MailboxProcessor<AsyncReplyChannel<'a>> = 
        MailboxProcessor.Start <| fun inbox ->
            let rec loop () = 
                async {
                    let! chn = inbox.Receive ()
                    chn.Reply current.Value
                    return! loop()
                }
            loop()

    // Serves incoming modify requests.
    let modifyProc : MailboxProcessor<('a -> 'a) * AsyncReplyChannel<UpdateResult>> = 
        MailboxProcessor.Start <| fun inbox ->
            let rec loop () = 
                async {
                    let! f, replyChannel = inbox.Receive()
                    let v = current.Value
                    try
                        current := f v
                        kernelDataChanged.Trigger current.Value
                        replyChannel.Reply UpdateResult.Ok
                    with
                    | e ->
                        replyChannel.Reply (UpdateResult.Error e.Message)

                    return! loop()
                }
            loop()
    {
        Current = fun () -> currentData.PostAndAsyncReply id
        KernelDataChanged = kernelDataChanged.Publish
        Modify = fun f -> modifyProc.PostAndAsyncReply(fun replyChannel -> f, replyChannel)
    }

// Default kernel state.
let defaultKernel = {
        Content = None
        Service = Local(localSvc<Document> (fun () -> { Content = Seq.empty; Format = { Width = 1m<m>; Height = 1m<m> } }) DocumentXml)
        Configuration = { ImplMe = false }
}